velox/common/caching/SsdFile.cpp

```diff
@@ -0,0 +1,412 @@
+#endif
+  fd_ = open(filename.c_str(), O_CREAT | O_RDWR | oDirect, S_IRUSR | S_IWUSR);
+  if (fd_ < 0) {
+    LOG(ERROR) << "Cannot open or create " << filename << " error " << errno;
+    exit(1);
+  }
+  readFile_ = std::make_unique<LocalReadFile>(fd_);
+  uint64_t size = lseek(fd_, 0, SEEK_END);
+  numRegions_ = size / kRegionSize;
+  fileSize_ = numRegions_ * kRegionSize;
+  if (size % kRegionSize > 0) {
+    ftruncate(fd_, fileSize_);
+  }
+  // The existing regions in the file are writable.
+  writableRegions_.resize(numRegions_);
+  std::iota(writableRegions_.begin(), writableRegions_.end(), 0);
+  regionScore_.resize(maxRegions_);
+  regionSize_.resize(maxRegions_);
+  regionPins_.resize(maxRegions_);
+}
+
+void SsdFile::pinRegion(uint64_t offset) {
+  std::lock_guard<std::mutex> l(mutex_);
+  pinRegionLocked(offset);
+}
+
+void SsdFile::unpinRegion(uint64_t offset) {
+  std::lock_guard<std::mutex> l(mutex_);
+  auto count = --regionPins_[regionIndex(offset)];
+  VELOX_CHECK_LE(0, count);
+  if (suspended_ && count == 0) {
+    growOrEvictLocked();
+  }
+}
+
+namespace {
+void addEntryToIovecs(AsyncDataCacheEntry& entry, std::vector<iovec>& iovecs) {
+  if (entry.tinyData()) {
+    iovecs.push_back({entry.tinyData(), static_cast<size_t>(entry.size())});
+    return;
+  }
+  auto& data = entry.data();
+  iovecs.reserve(iovecs.size() + data.numRuns());
+  int64_t bytesLeft = entry.size();
+  for (auto i = 0; i < data.numRuns(); ++i) {
+    auto run = data.runAt(i);
+    iovecs.push_back(
+        {run.data<char>(), std::min<size_t>(bytesLeft, run.numBytes())});
+    bytesLeft -= run.numBytes();
+    if (bytesLeft <= 0) {
+      break;
+    };
+  }
+}
+} // namespace
+
+SsdPin SsdFile::find(RawFileCacheKey key) {
+  SsdKey ssdKey{StringIdLease(fileIds(), key.fileNum), key.offset};
+  SsdRun run;
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    if (suspended_) {
+      return SsdPin();
+    }
+    newEventLocked();
+    auto it = entries_.find(ssdKey);
+    if (it == entries_.end()) {
+      return SsdPin();
+    }
+    run = it->second;
+    pinRegionLocked(run.offset());
+  }
+  return SsdPin(*this, run);
+}
+
+void SsdFile::newEventLocked() {
+  ++numEvents_;
+  if (numEvents_ > kDecayInterval && numEvents_ > entries_.size() / 2) {
+    numEvents_ = 0;
+    for (auto i = 0; i < numRegions_; ++i) {
+      int64_t score = regionScore_[i];
+      regionScore_[i] = (score * 15) / 16;
+    }
+  }
+}
+
+void SsdFile::load(SsdRun run, AsyncDataCacheEntry& entry) {
+  VELOX_CHECK_EQ(run.size(), entry.size());
+  regionScore_[regionIndex(run.offset())] += run.size();
+  ++stats_.entriesRead;
+  stats_.bytesRead += run.size();
+  if (entry.tinyData()) {
+    auto rc = pread(fd_, entry.tinyData(), run.size(), run.offset());
+    VELOX_CHECK_EQ(rc, run.size());
+  } else {
+    std::vector<struct iovec> iovecs;
+    addEntryToIovecs(entry, iovecs);
+
+    auto rc = folly::preadv(fd_, iovecs.data(), iovecs.size(), run.offset());
```

Comment:
SsdFile::read uses readFile_->preadv, but this code uses pread and folly::preadv. Why are these different? Would it make sense to make these two method use the same logic for reading from file?

Reply:
Looks like this comment got lost. Repeating. "SsdFile::read uses readFile_->preadv, but this code uses pread and folly::preadv. Why are these different? Would it make sense to make these two method use the same logic for reading from file?"